src/storage/src/render/upsert.rs

```diff
@@ -222,32 +295,104 @@ where
             // value2, time 7)
             let mut pending_values =
                 BTreeMap::<Timestamp, (Capability<Timestamp>, HashMap<_, UpsertSourceData>)>::new();
-            // this is a map of (decoded key) -> (decoded_value). We store the
-            // latest value for a given key that way we know what to retract if
-            // a new value with the same key comes along
-            let mut current_values = HashMap::new();
-
             // Intermediate structures re-used to limit allocations
             let mut scratch_vector = Vec::new();
+            let mut scratch_vector2 = Vec::new();
             let mut row_packer = mz_repr::Row::default();
             let mut dv = DatumVec::new();
 
-            // Prepare sorted and structured `key_indices` required
-            // by the upsert operator, and a `DatumVec` used to avoid
-            // an allocation.
-            let mut key_indices_sorted = upsert_envelope.key_indices.clone();
-            key_indices_sorted.sort_unstable();
             let key_indices_map = upsert_envelope
                 .key_indices
                 .iter()
                 .enumerate()
                 .map(|(idx, value_idx)| (*value_idx, idx))
                 .collect();
             let mut kdv = DatumVec::new();
+            // this is a map of (decoded key) -> (decoded_value). We store the
+            // latest value for a given key that way we know what to retract if
+            // a new value with the same key comes along
+            let mut current_values = None;
+
+            let mut initial_values_multiset: HashMap<(Row, Row), Diff> = HashMap::default();
```

Comment:
Do you possibly want a `ChangeBatch<(Row, Row)>` here instead? It's meant for doing the sort of thing you are doing (accumulating up modifications) and is probably higher throughput (at the expense of latency, which afaict you don't need).